### PR TITLE
More compression options

### LINKED ISSUE #3117: More compression options using CLI tools
## 🚀 Feature Request

The current checkpoint compression options are limited to those available in the python tarfile library and are not implemented very efficiently.

The current process is:

- an uncompressed file is saved to disk
- a temporary directory is created
- the file is copied to the temporary directory
- a compressed tar file is created from the temporary directory and written back to the original location

## Motivation

I did an investigation into the different compression options and found several interesting things:

- the choice of compression algorithm does not seem to matter much in terms of the final compression ratio (in my admittedly small sample). I think the checkpoints probably have an incompressible part and an easily compressible part so most algorithms end up compressing by similar amounts?
- the algorithms available in the `tarfile` library are slow compared to other options like `lz4`, `lzo` and `zstd`

Below are the results I gathered from implementing compression using CLI tools into the checkpoint
saving (PR #3118). The script used to benchmark is also attached.

| Format    | Compression Time (s) | Compressed size (MB) | Compressed size percent |
| --        | --                   | --                   | --                      |
| .pt       | 0.2                  | 283                  | 100.0%                  |
| .pt.bz2   | 13.6                 | 100                  | 35.3%                   |
| .pt.gz    | 5.9                  | 102                  | 36.0%                   |
| .pt.lz4   | 0.4                  | 117                  | 41.3%                   |
| .pt.lzma  | 43.7                 | 117                  | 41.3%                   |
| .pt.lzo   | 0.4                  | 117                  | 41.3%                   |
| .pt.xz    | 9.8                  | 97                   | 34.3%                   |
| .pt.zstd  | 0.5                  | 104                  | 36.7%                   |
| .tar      | 0.4                  | 283                  | 100.0%                  |
| .tar.bz2  | 13.7                 | 100                  | 35.3%                   |
| .tar.gz   | 7.7                  | 102                  | 36.0%                   |
| .tar.lzma | 41.3                 | 97                   | 34.3%                   |

So I do think that support for at least lz4, lzo and zstd are worth considering as they have very low overhead
but still provide significant size reduction in some cases.

The only downside is that not all systems will have all the CLI tools installed, so they would not be very good as the default setting but providing users with the option to use them could be useful.

[benchmark.zip](https://github.com/mosaicml/composer/files/14616928/benchmark.zip)

## Implementation

I have implemented support for this feature in #3118

### PR DESCRIPTION
**Pull request by mbway**
# What does this PR do?

This PR adds the ability to save compressed `.pt` checkpoint files using CLI compression tools available on the system (eg `.pt.lz4` using the `lz4` command).

See the related issue (#3117) for why this is useful. In summary, good compression is achievable with low overhead, but the system must have the required tools available so the default setting to not use compression remains.

## Potential Improvements
- more control over arguments passed to compressors (eg number of threads or compression level)
    - technically a user can modify `KNOWN_COMPRESSORS` and replace compressors with their own `CliCompressor` objects (which gives them full control over the arguments) but this is a bit hacky.
- ability to use CLI compressors with `.tar`, eg `.tar.lz4`

# What issue(s) does this change relate to?

Fixes #3117

# Before submitting
- [X] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [X] Did you update any related docs and document your change?
- [X] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [X] Did you run the tests locally to make sure they pass?
- [X] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Comment by mvpatel2000** [comment]
@mbway  feel free to ignore the docker failures -- unfortunately image building does not work from forks.

Could you please take a look at the CPU test failures though? Also you should be able to run this locally to ensure faster debugging.

I will make sure to review this soon as well in a first pass!

**Review by mvpatel2000 — COMMENTED** [review]
I think this mostly looks good to me besides test failures -- will provide a detailed pass once tests are clear.

@eracah can you also please look given checkpointing things?

**Comment by mvpatel2000** [comment]
Please feel free to re-request review from me when tests pass

File: composer/loggers/neptune_logger.py [hunk 1]
```diff
@@ -122,7 +122,7 @@ def neptune_run(self):
             self._neptune_run = Run(
                 project=self._project,
                 api_token=self._api_token,
-                mode=self._mode,
+                mode=self._mode,  # type: ignore
```
Comment (mvpatel2000) [inline composer/loggers/neptune_logger.py:125]:
```suggestion
                mode=self._mode,
```
This change seems spurious

File: composer/utils/checkpoint.py [hunk 2]
```diff
@@ -1143,18 +1144,29 @@ def _save_checkpoint(
...
 
-    with tempfile.TemporaryDirectory() as tmpdir:
-        shutil.move(filename, os.path.join(tmpdir, basename))
-        with tarfile.open(filename, write_mode) as tarball:
-            tarball.add(tmpdir, arcname='')
+            with tarfile.open(filename, write_mode) as tarball:
+                tarball.add(tmpdir, arcname='')
+
+    elif is_compressed_pt(filename):
+        log.debug('writing compressed checkpoint %s', filename)
```
Comment (mvpatel2000) [inline composer/utils/checkpoint.py:1161]:
```suggestion
        log.debug('Writing compressed checkpoint %s', filename)
```
nit

File: composer/utils/checkpoint.py [hunk 3]
```diff
@@ -1143,18 +1144,29 @@ def _save_checkpoint(
         return None
 
 
-def _compress_file(filename: str, basename: str):
-    """Replace a file with its compressed version.
+def _write_checkpoint_file(state_dict: Dict[str, Any], filename: str) -> None:
+    """Write the given checkpoint state to the given path. Compressing if indicated to do so by the file extension."""
+    if is_tar(filename):
+        log.debug('writing checkpoint tar file %s', filename)
```
Comment (mvpatel2000) [inline composer/utils/checkpoint.py:1150]:
```suggestion
        log.debug('Writing checkpoint tar file %s', filename)
```
nit

File: composer/utils/checkpoint.py [hunk 4]
```diff
@@ -1143,18 +1144,29 @@ def _save_checkpoint(
...
+                tarball.add(tmpdir, arcname='')
+
+    elif is_compressed_pt(filename):
+        log.debug('writing compressed checkpoint %s', filename)
+        compressor = get_compressor(filename)
+        with compressor.compress(filename) as f:
+            torch.save(state_dict, f)
+
+    else:
+        log.debug('writing uncompressed checkpoint %s', filename)
```
Comment (mvpatel2000) [inline composer/utils/checkpoint.py:1167]:
```suggestion
        log.debug('Writing uncompressed checkpoint %s', filename)
```
nit

File: tests/trainer/test_checkpoint.py [hunk 5]
```diff
@@ -29,7 +31,10 @@
 from composer.trainer import trainer
 from composer.trainer.trainer import Trainer
 from composer.utils import dist, is_tar, reproducibility
-from composer.utils.checkpoint import _ensure_valid_checkpoint, glob_filter
+from composer.utils.checkpoint import _ensure_valid_checkpoint  # type: ignore
+from composer.utils.checkpoint import _write_checkpoint_file  # type: ignore
+from composer.utils.checkpoint import _COMPOSER_STATES_FILENAME, glob_filter
```
Comment (mvpatel2000) [inline tests/trainer/test_checkpoint.py:36]:
```suggestion
from composer.utils.checkpoint import _ensure_valid_checkpoint, _COMPOSER_STATES_FILENAME, glob_filter, _write_checkpoint_file
```
nit: combine into one import

why is the type ignore needed? If it's unused, can we skip import?

**Review by mvpatel2000 — COMMENTED** [review]
This is a great PR! Thank you for the contribution.

This looks great to me. I left a few minor nits -- once those are addressed I can merge. If you have time to do it today, I can include in the release planned this week.

File: composer/loggers/neptune_logger.py (hunk 1 above)
Comment (mbway) [reply composer/loggers/neptune_logger.py:125]: when running the pre commit hooks, pyright actually gives lots of errors and warnings now. Originally there was just this one. I think I'm running the correct version of pyright:
```
composer/composer/callbacks/eval_output_logging_callback.py
  composer/composer/callbacks/eval_output_logging_callback.py:111:39 - error: Argument of type "list[str] | None" cannot be assigned to parameter "columns" of type "List[str]" in function "log_table"
    Type "list[str] | None" cannot be assigned to type "List[str]"
      Type "None" cannot be assigned to type "List[str]" (reportGeneralTypeIssues)
  composer/composer/callbacks/eval_output_logging_callback.py:111:64 - error: Argument of type "str | None" cannot be assigned to parameter "name" of type "str" in function "log_table"
    Type "str | None" cannot be assigned to type "str"
      Type "None" cannot be assigned to type "str" (reportGeneralTypeIssues)
composer/composer/utils/checkpoint.py
  composer/composer/utils/checkpoint.py:230:61 - error: Cannot access member "get_local_rank" for type "DeviceMesh"
    Member "get_local_rank" is unknown (reportGeneralTypeIssues)
  composer/composer/utils/checkpoint.py:283:56 - error: Cannot access member "get_group" for type "DeviceMesh"
    Member "get_group" is unknown (reportGeneralTypeIssues)
composer/composer/utils/object_store/sftp_object_store.py
  composer/composer/utils/object_store/sftp_object_store.py:21:12 - warning: Import "paramiko.client" could not be resolved from source (reportMissingModuleSource)
  composer/composer/utils/object_store/sftp_object_store.py:22:10 - warning: Import "paramiko" could not be resolved from source (reportMissingModuleSource)
  composer/composer/utils/object_store/sftp_object_store.py:190:14 - warning: Import "paramiko" could not be resolved from source (reportMissingModuleSource)
composer/tests/utils/object_store/object_store_settings.py
  composer/tests/utils/object_store/object_store_settings.py:45:12 - warning: Import "paramiko" could not be resolved from source (reportMissingModuleSource)
4 errors, 4 warnings, 0 informations

```

File: tests/trainer/test_checkpoint.py (hunk 5 above)
Comment (mbway) [reply tests/trainer/test_checkpoint.py:36]: more typechecking issues that may just be a problem with my setup.

**Review by mvpatel2000 — COMMENTED** [review]
Thanks! Force merging

**Comment by mvpatel2000** [comment]
Hm... it seems one of the tests fails and was not running as the previous testing image did not include the dependency:
https://github.com/mosaicml/composer/actions/runs/8383440850/job/22959154543

**Comment by mbway** [comment]
Sorry about that. The issue is caused by this bug https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=700681. The version of the `lzma` CLI tool that comes with `xz-utils` does not have this issue, so I think the fix is to remove `lzma` from the installation list in the dockerfile.

Also, the log highlights that `CliCompressor` should probably implement `__repr__` so it's easier to tell which compressor is the one causing the issue. I'll open a PR now.

**Pull request merged**